docs/usage/discovery-and-run.mdx

```diff
@@ -0,0 +1,158 @@
+---
+title: Discovery and Run
+---
+
+Pester runs your test files in two phases: Discovery and Run. During discovery, it quickly scans your test files and discovers all the Describes, Contexts, Its and other Pester blocks.
+
+This powers many of the features in Pester 5 and enables many others to be implemented in the future.
+
+To reap the benefits, there are two rules to follow:
+
+**Put all your code into `It`, `BeforeAll`, `BeforeEach`, `AfterAll` or `AfterEach`.**
+
+**Put no code directly into `Describe`, `Context` or on the top of your file, without wrapping it in one of these blocks.**
+
+All misplaced code will run during Discovery, and its results won't be available during Run. This will lead to confusing results. 
+
+
+## Execution order
+
+In Pester v4 the execution of your test code was mostly top-down, with the exception of all the `Before*` and `After*` blocks. Those blocks were picked up by Pester using AST and "moved" to the place where they needed to run. 
+
+For example placing an `AfterAll` at the top of a `Describe` would not run it at the top, but instead it would correctly run it as the last thing in that `Describe` wrapped in a `try finally`.
+
+Pester v5 takes this to the next level and manipulates the execution of your test file in a more complex way. This has good reason, but makes understanding how your code executes quite complex, especially when you don't follow the two rules listed above. 
+
+
+### Discovery
```

Comment:
I think it would be good to include in this section use of `BeforeDiscovery`. That is one thing I've found taking a bit to get used to how that is involved. 

Showing how this is involved in the Discovery phase and what effect it has on `Run` phase what be beneficial from my point of view.

Reply:
This is described in data driven tests.